onnxruntime/contrib_ops/webgpu/moe/qmoe.cc

```diff
@@ -0,0 +1,355 @@
+  const Tensor* fc2_experts_bias_optional = context.Input<Tensor>(7);
+  const Tensor* fc3_experts_weights_optional = context.Input<Tensor>(8);
+  const Tensor* fc3_scales_optional = context.Input<Tensor>(9);
+  const Tensor* fc3_experts_bias_optional = context.Input<Tensor>(10);
+  // zero points, not supported yet
+  const Tensor* fc1_zero_points = context.Input<Tensor>(11);
+  const Tensor* fc2_zero_points = context.Input<Tensor>(12);
+  const Tensor* fc3_zero_points = context.Input<Tensor>(13);
+
+  MoEParameters moe_params;
+
+  if (fc1_zero_points || fc2_zero_points || fc3_zero_points) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, NOT_IMPLEMENTED,
+                           "zero_points for QMoE are not yet supported on WebGPU.");
+  }
+
+  ORT_RETURN_IF_ERROR(::onnxruntime::contrib::moe_helper::CheckInputs<Tensor>(
+      moe_params, hidden_state, router_logits,
+      fc1_experts_weights, fc1_experts_bias_optional, fc1_scales, fc1_zero_points,
+      fc2_experts_weights, fc2_experts_bias_optional, fc2_scales, fc2_zero_points,
+      fc3_experts_weights_optional, fc3_experts_bias_optional, fc3_scales_optional, fc3_zero_points,
+      expert_weight_bits_ == 4 ? 2 : 1,
+      activation_type_ == MoEActivationType::SwiGLU, block_size_));
+
+  const auto& input_shape = hidden_state->Shape();
+
+  // SwiGLU validation
+  bool is_swiglu = (activation_type_ == MoEActivationType::SwiGLU);
+  if (fc3_experts_weights_optional) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, NOT_IMPLEMENTED,
+                           "FC3 gating is not yet implemented for non-SwiGLU activations on WebGPU.");
+  }
+
+  // process tokens in chunks of max_tokens to put some cap on memory usage
+  const int max_tokens = 512;
+
+  const uint32_t num_experts = static_cast<uint32_t>(moe_params.num_experts);
+  const uint32_t hidden_size = static_cast<uint32_t>(moe_params.hidden_size);
+  const int64_t fc1_output_size = is_swiglu && swiglu_fusion_ > 0 ? 2 * moe_params.inter_size : moe_params.inter_size;
+  const bool is_fp16 = hidden_state->DataType() == DataTypeImpl::GetType<MLFloat16>();
+  const auto dtype = is_fp16 ? DataTypeImpl::GetType<MLFloat16>() : DataTypeImpl::GetType<float>();
+  const auto dtype_uint32 = DataTypeImpl::GetType<uint32_t>();
+
+  const int64_t K_fc1 = moe_params.hidden_size;
+  const int64_t N_fc1 = fc1_output_size;
+  const int64_t K_fc2 = moe_params.inter_size;
+  const int64_t N_fc2 = moe_params.inter_size;
+  const int64_t accuracy_level = 4;
+  const int64_t block_size_fc1 = (block_size_ != 0) ? block_size_ : K_fc1;
+  const int64_t block_size_fc2 = (block_size_ != 0) ? block_size_ : K_fc2;
+  Status status;
+
+  Tensor* output_tensor = context.Output(0, input_shape);
+  const int total_output_size = (static_cast<int>(input_shape.Size()) + 3) / 4;
+
+  // we are accumulating expert results into output_tensor, need to initialize to zero
+  ZeroTensorProgram zero;
+  zero
+      .AddOutput({output_tensor, ProgramTensorMetadataDependency::None, 4})
```

Comment:
You reverted below change. Why? `ProgramTensorMetadataDependency::None` is not correct. Otherwise, how do you distinguish f16 and f32. And using flatten can make sure the correctness even your last dimension is not divisible by 4.
```suggestion
      .AddOutput({output_tensor, ProgramTensorMetadataDependency::Type, ProgramInput::Flatten, 4})
```

Reply:
done